File: Accuracer/Demos/Delphi/ResturctureMultiThread/uBkThread.pas
unit uBkThread;

interface

uses Classes, SysUtils, Windows, ACRMain, ACRTypes;


type
  TRestructureThread = class (TThread)
    private
     FDBFileName:    String;
     FTableName:     String;
    protected
     procedure Finalize;
    public
     constructor Create(aDBFile, aTableName: String);
     procedure Execute; override;
  end;

implementation

uses uMain;

{ TRestructureThread }

constructor TRestructureThread.Create(aDBFile, aTableName: String);
begin
  inherited Create(True);
  FDBFileName := aDBFile;
  FTableName := aTableName;
end;

procedure TRestructureThread.Execute;
var
    db: TACRDatabase;
    t:  TACRTable;
    i:  Integer;
begin
  db := TACRDatabase.Create(nil);
  t := TACRTable.Create(nil);
  try
    db.DatabaseFileName := FDBFileName;
    db.Open;
    t.TableName := FTableName;
    t.DatabaseName := db.DatabaseName;
    t.Open;
    t.Close;
    i := t.RestructureFieldDefs.Count;
    t.RestructureFieldDefs.Add('Field #'+IntToStr(i),aftChar,20);
    t.RestructureTable;
  finally
    t.Free;
    db.Free;
    Synchronize(Finalize);
  end;
end;

procedure TRestructureThread.Finalize;
begin
  Form1.Finish;
end;

end.

File: Accuracer/Demos/C++ Builder 5/ResturctureMultiThread/uBkThread.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "uBkThread.h"
#include "uMain.h"

void __fastcall TRestructureThread::Finish()
{
  Form1->Finish();
}

void __fastcall TRestructureThread::Finalize()
{
  TThreadMethod tm = &Finish;
  Synchronize(tm);
}

__fastcall TRestructureThread::TRestructureThread(WideString aDBFile, WideString aTableName, bool CreateSuspended)
: TThread(CreateSuspended)
{
 FTableName = aTableName;
 FDBFileName = aDBFile;
}

void __fastcall TRestructureThread::Execute()
{
  TACRDatabase *db;
  TACRTable *t;
  int i;
  db = new TACRDatabase(NULL);
  t = new TACRTable(NULL);
  try
  {
    db->DatabaseFileName = FDBFileName;
    db->Open();
    t->TableName = FTableName;
    t->DatabaseName = db->DatabaseName;
    t->Open();
    t->Close();
    i = t->RestructureFieldDefs->Count;
    t->RestructureFieldDefs->Add("Field #"+IntToStr(i),aftChar,20);
    t->RestructureTable();
  }
  __finally
  {
    delete t;
    delete db;
    Finalize();
  };
}


//---------------------------------------------------------------------------

#pragma package(smart_init)
